storage/badger/operation/reader_batch_writer.go

```diff
@@ -0,0 +1,130 @@
+package operation
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/dgraph-io/badger/v2"
+
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/storage"
+)
+
+type ReaderBatchWriter struct {
+	db    *badger.DB
+	batch *badger.WriteBatch
+
+	addingCallback sync.Mutex // protect callbacks
+	callbacks      []func(error)
+}
+
+var _ storage.BadgerReaderBatchWriter = (*ReaderBatchWriter)(nil)
+
+func (b *ReaderBatchWriter) ReaderWriter() (storage.Reader, storage.Writer) {
```

Comment:
Makes sense. 

I originally combined them because there is a few places need intermingling reads. But I think we probably are able to refactor those logic, such as bootstrapping a sealing segment.